src/security/intel/cbnt/measurement.c

```diff
@@ -0,0 +1,543 @@
+#include <device/mmio.h>
+#include <intelblocks/cse.h>
+#include <security/intel/cbnt/cbnt.h>
+#include <security/tpm/tspi/crtm.h>
+#include <security/vboot/misc.h>
+
+#define TPM_ALG_RSA     0x0001
+#define TPM_ALG_ECC     0x0023
+
+#define TPM_ALG_RSASSA  0x0014
+#define TPM_ALG_RSAPSS  0x0016
+#define TPM_ALG_ECDSA   0x0018
+#define TPM_ALG_SM2     0x001b
+
+/* Bits of `struct bpm_ibbs::flags`. */
+#define BPM_IBBS_FLAG_RESERVED1             (1 << 0) /* must be set to 1 */
+#define BPM_IBBS_FLAG_RESERVED2             (1 << 1) /* must be set to 1 */
+#define BPM_IBBS_FLAG_AUTH_MEASURE          (1 << 2) /* extend PCR-7 */
+#define BPM_IBBS_FLAG_CAP_PCRS_ON_TPM_FAIL  (1 << 3)
+#define BPM_IBBS_FLAG_TOP_SWAP_SUPPORTED    (1 << 4)
+#define BPM_IBBS_FLAG_FORCE_TME             (1 << 5)
+#define BPM_IBBS_FLAG_FORCE_IGN             (1 << 6)
+#define BPM_IBBS_FLAG_SRTM_AC               (1 << 7)
+
+/* Bits of `struct km_hash::usage` indicating to which public key the hash corresponds. */
+#define KM_HASH_USAGE_BPM       (1 << 0)
+#define KM_HASH_USAGE_FIT       (1 << 1)
+#define KM_HASH_USAGE_ACM       (1 << 2)
+#define KM_HASH_USAGE_SDEV      (1 << 3)
+#define KM_HASH_USAGE_PFR_CPLD  (1 << 4)
+
+/* These names are hard-coded in ./Makefile. */
+#define INTEL_CBNT_CBFS_KM   "key_manifest.bin"
+#define INTEL_CBNT_CBFS_BPM  "boot_policy_manifest.bin"
+
+/* Contains SHA256 hash of OEM's public key. */
+#define OEM_CRED_FILE  "/fpf/intel/OemCred"
+
+/*
+ * Definitions of the structures come from Intel document #575623 CBnT BWG v1.2.5 ("Intel
+ * Converged Boot Guard and Intel Trusted Execution Technology (Intel TXT) BIOS Specification").
+ * It is specific for TigerLake and newer platforms.
+ *
+ * Versions in structures could be different for other revisions of the document.
+ */
+
+/* Header of the Boot Policy Manifest. */
+struct __packed bpm_header {
+	char structure_id[8];     /* "__ACBP__" */
+	uint8_t struct_ver;       /* 0x24 */
+	uint8_t hdr_struct_ver;   /* 0x20 */
+	uint16_t hdr_size;
+	uint16_t key_sig_offset;  /* offset to KeySignature field of Policy Manifest Signature
+				     Element (PMSE) */
+	uint8_t bpm_revision;
+	uint8_t bpm_svn;
+	uint8_t acmsvn_auth;
+	uint8_t reserved;
+	uint16_t nem_data_stack;  /* in 4 KiB pages */
+	uint8_t se_element[];
+};
+
+/* Deprecated uses must be initialized as { .alg = TPM_ALG_NULL (0x10), .size = 0 } */
+struct __packed hash_struct {
+	uint16_t alg;
+	uint16_t size;
+	uint8_t data[];
+};
+
+struct __packed bpm_hash_list {
+	uint16_t size;
+	uint16_t count;
+	struct hash_struct hashes[];
+};
+
+/* IBB Segment Element (upper part) */
+struct __packed bpm_ibbs {
+	char structure_id[8];  /* "__IBBS__" */
+	uint8_t struct_ver;    /* 0x21 */
+	uint8_t reserved1;
+	uint16_t element_size;
+	uint8_t reserved2;
+	uint8_t set_number;    /* 0 */
+	uint8_t reserved3;
+	uint8_t pbet_value;
+	uint32_t flags;        /* see BPM_IBBS_FLAG_* constants for values */
+	uint64_t bar0;
+	uint64_t bar1;
```

Comment:
These are VTD/IOMMU BARs.
```suggestion
	uint64_t iommu_bar0;
	uint64_t iommu_bar1;
```

Reply:
Renamed: https://github.com/Dasharo/coreboot/compare/ffa516f0285a87e049c9e7ebefc0b23e3d941a77..4b501619643dc15f6eb104ad64919f0fb0b5e347#diff-1725d4cd7234cc446584ac6ebad791ed8c8e640970d4adb2e99e5b5b233030f7R86